### PR TITLE
CFG generation throws StackOverlowException when syntax tree is too deep

### LINKED ISSUE #3630: CFG generation throws StackOverlowException when syntax tree is too deep
### Description

CFG generation can throw `StackOverlofException` that will break the build of analyzed project.

Project `Lucene.NET` has [a funny file](https://github.com/apache/lucenenet/blob/master/src/Lucene.Net.Analysis.Common/Analysis/CharFilter/HTMLStripCharFilter.cs) with constants initialized to 1800 string concatenations. When we generate CFG for such syntax node with a very deep structure, it fails on `StackOverflowException`. 

Same problem occurs for all `BinaryExpressionSyntax` (&&, ||, ...). Maximum depth varies on .NET Framework vs .NET Core build.
### Reproducer

Issue is reproducible with simple test:
```C#
[TestMethod]
public void CauseStackOverflow()
{
    const int count = 2000;
    const string dna = "ACGT";
    var concatenations = Enumerable.Repeat($@"""{dna}""", count).JoinStr(" +\n");
    var cfg = Build($"var x = {concatenations};");
    VerifyMinimalCfg(cfg);
    VerifyAllInstructions(cfg.EntryBlock);
}
```

Note: `StackOverflowException` cannot be caught. Failure of the test is visible in Test output window.
### Possible solutions

1. Use Roslyn CFG
2. Validate and limit nesting depth of supported CFG
3. Process BinaryExpressionSyntax in non-recursive way
4. Do both 2. and 3.

### After it's fixed

Once this is fixed, we need to:
* Create issue on SONARSEC side to update `SonarAnalyzer.CFG.dll` version to the latest released
* Remove `SupportedNodeCountLimit` from Sonar-Securify C# FE `ConstructorUcfgBuilder.ProcessInitializerValue`

### PR DESCRIPTION
**Pull request by pavel-mikula-sonarsource**
Fixes #3630

### CODE REVIEW

File: analyzers/src/SonarAnalyzer.CFG/ControlFlowGraph/CSharpControlFlowGraphBuilder.cs [hunk 1]
```diff
@@ -528,6 +534,12 @@ private Block BuildPatternExpression(PatternSyntaxWrapper patternSyntaxWrapper,
             throw new NotSupportedException($"{patternSyntaxWrapper.SyntaxNode.Kind()}");
         }
 
+        private static bool IsTooComplex(SyntaxNode node)
```
Comment (andrei-epure-sonarsource) [inline analyzers/src/SonarAnalyzer.CFG/ControlFlowGraph/CSharpControlFlowGraphBuilder.cs:537]: given you only use it in `BuildExpression`, can't this be a static local function?

(this class already has lots of methods - and you group this private static method inside the region `Top level Build` - you may as well keep it as a local function)

File: analyzers/tests/SonarAnalyzer.UnitTest/ControlFlowGraph/ControlFlowGraphTest.cs [hunk 2]
```diff
@@ -177,6 +173,52 @@ public class Bar
             VerifyAllInstructions(exit);
         }
 
+        [TestMethod]
+        [TestCategory("CFG")]
+        public void Cfg_ExtremelyNestedExpression_NotSupported_FromExpression()
```
Comment (andrei-epure-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/ControlFlowGraph/ControlFlowGraphTest.cs:178]: both `Cfg_ExtremelyNestedExpression_NotSupported_FromExpression` and `Cfg_ExtremelyNestedExpression_NotSupported_FromBodyMethod` are from body methods...

You could rename 
```suggestion
        public void Cfg_ExtremelyNestedExpression_NotSupported_FromAssignment()
```

File: analyzers/tests/SonarAnalyzer.UnitTest/ControlFlowGraph/ControlFlowGraphTest.cs [hunk 3]
```diff
@@ -177,6 +173,52 @@ public class Bar
...
+            var method = CompileWithMethodBody(string.Format(TestInput, $"var x = {ExtremelyNestedExpression()};"), "Bar", out var semanticModel);
+            Action a = () => CSharpControlFlowGraph.Create(method.Body, semanticModel);
+
+            a.Should().Throw<NotSupportedException>().WithMessage("Too complex expression");
+            CSharpControlFlowGraph.TryGet(method.Body, semanticModel, out _).Should().BeFalse();
+        }
+
+        [TestMethod]
+        [TestCategory("CFG")]
+        public void Cfg_ExtremelyNestedExpression_NotSupported_FromBodyMethod()
```
Comment (andrei-epure-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/ControlFlowGraph/ControlFlowGraphTest.cs:189]:
```suggestion
        public void Cfg_ExtremelyNestedExpression_NotSupported_FromBodyMethodReturn()
```

**Review by andrei-epure-sonarsource — CHANGES_REQUESTED** [review]
Do you think it's worth adding an integration test as well? (because we don't test what's the end result when throwing the exception)

File: analyzers/src/SonarAnalyzer.CFG/ControlFlowGraph/CSharpControlFlowGraphBuilder.cs (hunk 1 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/src/SonarAnalyzer.CFG/ControlFlowGraph/CSharpControlFlowGraphBuilder.cs:537]: Method `BuildExpression` itself has almost 300 LOC. I don't want to add local function there.

File: analyzers/tests/SonarAnalyzer.UnitTest/ControlFlowGraph/ControlFlowGraphTest.cs (hunk 2 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/ControlFlowGraph/ControlFlowGraphTest.cs:178]: Good catch! I actually wanted to check `FromExpression`. My intention is to cover all branches here:
https://github.com/SonarSource/sonar-dotnet/blob/84c8cde333395a1f13e27f95832b9e1937c8721c/analyzers/src/SonarAnalyzer.CFG/ControlFlowGraph/CSharpControlFlowGraphBuilder.cs#L79-L90

File: analyzers/tests/SonarAnalyzer.UnitTest/ControlFlowGraph/ControlFlowGraphTest.cs (hunk 3 above)
Comment (pavel-mikula-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/ControlFlowGraph/ControlFlowGraphTest.cs:189]: `FromArrowMethod` and `FromBodyMethod` makes a symmetic pair in the list.

**Comment by pavel-mikula-sonarsource** [comment]
I don't think adding integration test is needed. The UTs are testing the `TryGet` method to return `null`.

**Review by andrei-epure-sonarsource — APPROVED** [review]
LGTM

**Pull request merged**